Expression executor service

flash1293 opened the pull request:
## Summary
Adds a simple version of the expression executor service to the data plugin because this functionality will be used in a wide variety of plugins and apps in Kibana.

## Context
This will be used in Lens to implement https://github.com/elastic/kibana/issues/36187
Over time this service is expected to grow (hopefully in a backwards compatible manner) into all the things outlined here https://github.com/elastic/kibana/issues/25374

The first simple API `run(expression, domElement)` will become a convenience function on top of these functions. It only provides a really simple API to run and render given expressions. Comments in the code mark unfinished parts which are not necessary to get this basic functionality.

## Integration
This is part of the data plugin which is currently a shimmed NP plugin. To get as close as possible to the NP way of doing things, the stateful dependencies (getInterpreter and renderersRegistry) are passed down from the main plugin entry point `src/legacy/core_plugins/data/public/index.ts`.

As types are not available yet because the interpreter plugin itself is still JS, this integration uses local types for now. These are expected to be removed once the interpreter plugin is moved to TS.

elasticmachine commented:
Pinging @elastic/kibana-app-arch

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14656/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14703/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14955/)

ppisljar commented on src/legacy/core_plugins/data/public/expression_executor/expression_executor_service.tsx:
```diff
@@ -0,0 +1,177 @@
+       * This is a React bridge of the `run` method
+       * @param props
+       */
+      ExpressionRenderer({ expression }: { expression: string }) {
```
nit: extract to a separate file

ppisljar commented on src/legacy/core_plugins/data/public/expression_executor/expression_executor_service.tsx:
```diff
@@ -0,0 +1,177 @@
+     * @param element
+     */
+    const run = async (expressionOrAst: string | Ast, element: Element) => {
+      if (!this.interpreterInstance) {
```
create instance in the setup directly ?

ppisljar commented on src/legacy/core_plugins/data/public/index.ts:
```diff
@@ -25,25 +34,34 @@ class DataPlugin {
       indexPatterns: this.indexPatterns.setup(),
       search: this.search.setup(),
       query: this.query.setup(),
+      expressionExecutor: this.expressionExecutor.setup(null, {
```
lets name it interpreter service

flash1293 replied:
Sounds good - this means we will merge it with the current interpreter plugin, right?

ppisljar replied:
++
`data.expressions.fromAst()` and `data.expression.toAst()` probably make more sense then
`data.expressions.run()` is also fine and `data.expressions.ExpressionRenderer` component
it would be nice if we could have `data.expressions.interpreter` instead of `data.expressions.getInterpreter()`
and then the registries:
`data.expressions.functionRegistry`, `data.expressions.renderersRegistry`, `data.expressions.typesRegistry`

ppisljar commented on src/legacy/core_plugins/data/public/index.ts:
```diff
@@ -25,25 +34,34 @@ class DataPlugin {
       query: this.query.setup(),
+      expressionExecutor: this.expressionExecutor.setup(null, {
+        interpreter: {
+          getInterpreter,
```
i think its ok to import these inside the interpreter service directly @lukeelmers  what do you think ?
as interpreter plugin will move in here as well (so this is pretty much the doing re-exporting we are doing in other data services)

flash1293 replied:
Talked with @ppisljar about this, currently the interpreter imports fail when not done on the top level. I will leave it like this for now and it can be removed when moving interpreter over

ppisljar commented on src/legacy/core_plugins/data/public/expression_executor/expression_executor_service.tsx:
```diff
@@ -0,0 +1,177 @@
+
+import { CoreSetup } from 'kibana/public';
+import { useRef, useEffect } from 'react';
+import React from 'react';
```
react shouldn't be part of the API, we can have react components wrapping the api but pls put in separate file

flash1293 replied:
So we don't do stateful components at all? Or is this just about the React import in the service itself?

The nice thing about delivering components like that is that they are usable directly (`<ExpressionRenderer expression="..." />`) instead of having to pass the stateful thing in (e.g. `<ExpressionRenderer runFn={plugins.expressionExecutor.run} expression="..." />`)

lukeelmers replied:
Yeah agreed with @ppisljar here -- I'd suggest providing a `run` function to just run the expression, and a in a separate file a "convenience" method for React, which uses `run` under the hood.

Then both can be returned from `setup` as part of the public contract. This way people have the option of using the React wrapper or the `run` method directly. Also some people might want to run an expression and get the output without having to use a renderer.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14968/)

flash1293 commented on src/legacy/core_plugins/data/public/index.ts:
```diff
@@ -57,9 +75,12 @@ export const data = new DataPlugin().setup();
 }
 
+export { ExpressionExecutorSetup } from './expression_executor';
+
```
I don't necessarily _need_ it by itself, I just planned to save the instance of this service in my consumer plugin like `private expressionsService: ExpressionsServiceSetup;`.
If this isn't exported directly I have to plug it out of the DataSetup like `private expressionsService: DataSetup['expressionsService'];`.

But that would also work, I very happy to follow best practices here. Reading your other comments, these best practices are importing only the leaf types which are actually passed around, so I will change it to that (the react component and its props plus the signature of the "raw" run function).

flash1293 commented on src/legacy/core_plugins/data/public/expression_executor/expression_executor_service.tsx:
```diff
@@ -0,0 +1,177 @@
+  onDestroy: (fn: () => void) => void;
+}
+
+export interface RenderFunction {
```
Makes sense, thanks. These types are only exported for the tests, I changed my code to re-export public stuff in index

flash1293 commented on src/legacy/core_plugins/data/public/expression_executor/expression_executor_service.tsx:
```diff
@@ -0,0 +1,177 @@
+export class ExpressionExecutorService {
+  private interpreterInstance: Interpreter | null = null;
+
+  // TODO core won't ever be null once this is switched to the new platform
```
This was a misunderstanding from my side, sorry. I thought the individual services were planned to become plugins in the future because of the very similar call signature. I removed the core parameter completely, this also should make clear the service isn't depending on it

flash1293 commented on src/legacy/ui/public/visualize/loader/pipeline_helpers/run_pipeline.ts:
```diff
@@ -25,15 +25,15 @@ import { getInterpreter } from 'plugins/interpreter/interpreter';
 import { Filters, Query, TimeRange } from 'ui/visualize';
 
-interface InitialContextObject {
+export interface InitialContextObject {
```
They are used in the expressions service inside `data` to help with the local interim types there because atm there are not other types available. +1 for more complete expression types. As soon as your PR is merged I will fix the type references in the expressions service.

flash1293 commented:
@ppisljar @lukeelmers Thanks for your review, I think I addressed all the points.

About naming: I renamed the service to "ExpressionsService", the functions are still called `run` and `ExpressionRenderer`. I have no strong opinion about them - if you come up with better names please let me know. It should somehow say "Interpret and execute an expression and render the result if a separate dom element was also passed to the function". `run` might be too generic, not sure here. I know about the previous discussion of passing just the expression and getting a `Handler` object which does the other things, but I would veto doing this now, this function is meant to be a shortcut for that multi step process.

ppisljar commented on src/legacy/core_plugins/data/public/expressions/expressions_service.ts:
```diff
@@ -0,0 +1,113 @@
+    interpreter: { renderersRegistry, getInterpreter },
+  }: ExpressionsServiceDependencies) {
+    const run = createRunFn(
+      renderersRegistry,
```
we should also expose `interpreter` and `renderersRegistry` but no need to do it in this PR.

ppisljar commented on src/legacy/core_plugins/data/public/index.ts:
```diff
@@ -25,25 +34,34 @@ class DataPlugin {
   }
 
-  public setup() {
+  public setup(): DataSetup {
```
@lukeelmers could setup function be async ? then we could expose interpreter in our contract instead of promise which resolves with interpreter.

streamich replied:
It can be async, but it will slow down loading of all plugins.

ppisljar replied:
lets leave it for now, most likely we can refactor interpreter to not need async initialization. atm it only needs it to get the list of server functions.

flash1293 replied:
Left a TODO in the code

ppisljar commented on src/legacy/core_plugins/data/public/expressions/expression_runner.ts:
```diff
@@ -0,0 +1,65 @@
+    ast,
+    { type: 'null' },
+    {
+      getInitialContext: () => ({}),
```
this should be one of the properties to the function, whoever runs the expression might want to provide initial context.

ppisljar commented on src/legacy/core_plugins/data/public/expressions/expression_runner.ts:
```diff
@@ -0,0 +1,65 @@
+    typeof expressionOrAst === 'string' ? fromExpression(expressionOrAst) : expressionOrAst;
+  const response = await interpreter.interpretAst(
+    ast,
+    { type: 'null' },
```
context should be one of the properties, needs to be provided from outside and we should default to null.

ppisljar commented on src/legacy/core_plugins/data/public/expressions/expressions_service.ts:
```diff
@@ -0,0 +1,113 @@
+// TODO:
+// this type import and the types below them should be switched to the types of
+// the interpreter plugin itself once they are ready
+import { RunPipelineHandlers } from 'ui/visualize/loader/pipeline_helpers/run_pipeline';
```
nit: can you for now just duplicate that code here, i will be removing run_pipeline helper as soon as this gets merged

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15203/)

flash1293 commented:
@ppisljar @lukeelmers I added an options parameter for context and getInitialContext if you want to take a final look.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15230/)

lizozom commented:
@flash1293 Will this be backported to 7.x?
